### PR TITLE
Add tutorial basics

### PR DESCRIPTION
**Pull request by reichlfl**
**Tutorial basics and resources**

_Note: Do not review this commit-by-commit, it contains a final "cleanup" commit that removes all intermediate changes done to the application._

Added a framework to create interactive tutorials (tutorialoverlay.h). Documentation on how to create new tutorials can be found there, and in the accompanying UI file (tutorialoverlay.ui).

Also added a bunch of resources used by the tutorial system, as well as some yet unused but potentially usefull.

Note that the design is still preliminary, we probably want to change that in the future - especially the border around the highlighted widget isn't perfect.

**Examples**

How it looks in Orbit:
image

Creating new steps in the QT Designer:
image

### CODE REVIEW

**Comment by beckerhe** [comment]
I haven't looked at the code yet, just some comments regarding the font files upfront:

- To me it looks like we're both embedding the font-files into Orbit and distributing them as separate files. I guess we shouldn't do both.
- Also, I think we need a license file for the font-files and according to Google rules they have to be placed in a subdirectory under third_party/ since we are not the copyright owner.
- Furthermore, if I understood the OFL correctly we have to distribute the font-files even if we decide to embed the fonts.

I have said that before, but I think it's worth mentioning again: This is super cool! :smiley:

File: OrbitQt/TutorialOverlay.h
Comment (pierricgimmig) [inline OrbitQt/TutorialOverlay.h:126]: Storing the iterator seems dangerous to me. The iterator will become invalid on vector reallocation. Could active_step be an index instead? Maybe we should encapsulate those fields also and provide "IncrementActiveStep" and "DecrementActiveStep" methods?

File: OrbitCore/Path.h
Comment (reichlfl) [reply OrbitCore/Path.h:30]: This kicked off something bigger :D I've added [[nodiscard]] to the remaining methods as well for consistency, which resulted in this: https://github.com/google/orbit/pull/1130

I'll rebase this PR as soon as it's ready to be merged. Thanks for pointing it out!

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp:126]: Maybe the name "verify" is a bit misleading, since it's more meant to check if anything is preventing us from ending the current step. The logic here is "if there is no active step, nothing prevents us from going further".

But I've realized that `EndActiveStep(true)` is only used in one place anyways, so I've removed the parameter and call `VerifyActiveStep` now explicitly in this one location. I believe that makes it more readable, thanks!

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h:155]: See comment above, I've removed the parameter

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Lesson learned: If I already think it's a bit ugly and requires a comment to explain why I'm doing this, it will not make it through code review ;-)

You're absolutely right, I changed this in a slightly different way: Added a `StepExists` method that is used here and in `GetStep` to unify this, and added a test for it.

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Right, I forgot that the iterator will become invalid in this case. It can't really happen because the API only allows you to add a complete section at once, so `step_names` will never be modified. Nevertheless I can see this becoming a problem if we decide to add some functionality on section management later, so I changed that.

I didn't add the Increment/Decrement methods though, because this is kind of what the PrevStep / NextStep is for - granted, they do a bit more, but otherwise I'd need to duplicate some logic or prepare the Increment / Decrement in a way that they're only used by Prev/NextStep anyways...

**Comment by reichlfl** [comment]
@pierricgimmig I've updated the PR to reflect your discussion items

File: OrbitQt/CutoutWidget.h
Comment (beckerhe) [inline OrbitQt/CutoutWidget.h:10]: A short comment stating the use-case of this widget and that it is related to the TutorialOverlay would be great.

File: OrbitQt/TutorialOverlay.h
Comment (beckerhe) [inline OrbitQt/TutorialOverlay.h:101]: Did you mean `const std::string&`?

File: OrbitQt/main.cpp
Comment (beckerhe) [inline OrbitQt/main.cpp:198]: Can you replace std::string::c_str() by QString::fromStdString?

**Comment by beckerhe** [comment]
Sorry for being late. I added some comments.

File: OrbitQt/TutorialOverlay.cpp
Comment (pierricgimmig) [inline OrbitQt/TutorialOverlay.cpp:173]: Shouldn't we check for "section.active_step_index >= section.step_names.size()" ?

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Done

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Done

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Done

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Done

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Ha, I was looking for something like that, thanks!

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Done

File: OrbitQt/CutoutWidget.h
Comment (reichlfl) [reply OrbitQt/CutoutWidget.h:21]: Good point, done.

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Indeed I did

File: OrbitQt/main.cpp
Comment (reichlfl) [reply OrbitQt/main.cpp]: Done

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Right, thanks. In fact it didn't make a difference because `HasActiveStep()` returns false when `active_step_index >= step_names.size()`, but still this isn't what I meant to achieve here.

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h:160]: This can't be static because it stops to recurse if it encounters TutorialOverlay's own parent.

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h]: Done

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h:132]: Using unordered_map now since flat_hash_map behaves a bit differently regarding iterator validity, and this causes issues

File: OrbitQt/TutorialOverlay.h
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.h:133]: Done

File: OrbitQt/TutorialOverlay.cpp
Comment (reichlfl) [reply OrbitQt/TutorialOverlay.cpp]: Changed to const std::string&

**Comment by reichlfl** [comment]
> I haven't looked at the code yet, just some comments regarding the font files upfront:
> 
> * To me it looks like we're both embedding the font-files into Orbit and distributing them as separate files. I guess we shouldn't do both.
> * Also, I think we need a license file for the font-files and according to Google rules they have to be placed in a subdirectory under third_party/ since we are not the copyright owner.

I've added the license file and embedded all icons and fonts now. I've kept the existing structure for now since it was like that in the past, we should decide if we move everything into a different structure (including the icons).

> * Furthermore, if I understood the OFL correctly we have to distribute the font-files even if we decide to embed the fonts.

Where does it say that? The way I read it is distributed an embedded version is fine, it just needs to contain the license.

**Comment by reichlfl** [comment]
@beckerhe @pierricgimmig All comments should be adressed now. I'll wait with merging this until after the branch cut anyways, but could you re-check and approve if you have no further comments? Thanks!

**Comment by beckerhe** [comment]
> > I haven't looked at the code yet, just some comments regarding the font files upfront:
> > 
> > * To me it looks like we're both embedding the font-files into Orbit and distributing them as separate files. I guess we shouldn't do both.
> > * Also, I think we need a license file for the font-files and according to Google rules they have to be placed in a subdirectory under third_party/ since we are not the copyright owner.
> 
> I've added the license file and embedded all icons and fonts now. I've kept the existing structure for now since it was like that in the past, we should decide if we move everything into a different structure (including the icons).
> 
> > * Furthermore, if I understood the OFL correctly we have to distribute the font-files even if we decide to embed the fonts.
> 
> Where does it say that? The way I read it is distributed an embedded version is fine, it just needs to contain the license.

Yes that might be the case. I'm really not an expert on this. As far as I understood, embedding is only when it's embedded in a document, but I might be wrong here. Anyway, when it's "embedding" we are fine. When it's "distribution" we have to offer a way to extract the font from our bundle or provide a copy of the font. (https://scripts.sil.org/cms/scripts/page.php?item_id=OFL-FAQ_web#fa29573b) But I might be even wrong about that. Pretty difficult the license stuff. This has to be approvide by Google's OpenSource Licensing group anyway, so maybe we just go ahead and see what they say.

**Review by beckerhe — APPROVED** [review]
Thanks a lot for that really nice feature!

**Comment by reichlfl** [comment]
FYI, I removed the font and iconography which requires additional licenses from this PR for now as we are in discussion on how to style the tutorials anyways. I'm merging with a more neutral look now and will update as soon as we agreed on this and started integrating the first tutorials for the users.

**Pull request merged**
